Make the native build oblivious to Microsoft RN vs patched Facebook RN

NickGerleman opened the pull request:
Extracted from https://github.com/microsoft/react-native-windows/pull/3941

This change is meant to reduce the last of the differences between how the native build treats facebook/react-native and microsoft/react-native. These changes center around a few issues.

facebook/react-native with source patching can now match the DLL exports of micosoft/react-native. Update these so that we will be able to use all previously exported functions. Make unpatched react native the new "main" export file to try to prevent future breaks of people keeping them out of sync. Remove "x86sdx" deps which we don't seem to build, or textually reference anywhere in the project. 

We previously did not disable RTTI for facebook/react-native builds of ReactUwp. There weren't any diffs in the current version of ReactCommon that would seem to necessitate this change, and we end up building correctly without it. Having per-project opting into disabling RTTI itself seems hairy. We have enough shared code between projects that do and don't disable  RTTI, that we would likely not be able to use it in most places. This motivates disabling RTTI for all projects to decrease binary cost. External projects generated using the react-native-cli vnext template are not subject to this change since they do not import React.Cpp.props.

We can now remove all build logic around the "OSS_RN" property, historically used to differentiate between builds using upstream React Native. 

Remove some historical but now unused build logic on the way. 

This should work with current CI, but will leave us in a brief window where we
run CI on upstream react-native with patches, but not without. This change has been
validated alongside a CI change which does however.

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/microsoft/react-native-windows/pull/3948)

NickGerleman commented on vnext/ReactUWP/ReactUWP.vcxproj:
```diff
@@ -115,55 +115,38 @@
   </ItemDefinitionGroup>
-  <ItemDefinitionGroup Condition="'$(Platform)'=='ARM' AND '$(OSS_RN)'=='true'">
-    <Link>
-      <ModuleDefinitionFile>Endpoints\dll\react-native-uwp.oss.arm.def</ModuleDefinitionFile>
```
ARM exposed less in its defs file, so we can share between patched and unpatched. It's not ideal that the exported functions weren't consistent, but it seems better to keep the less exposed ABI surface as is.

NickGerleman commented on vnext/ReactUWP/ReactUWP.vcxproj:
```diff
@@ -115,55 +115,38 @@
+
+  <ItemDefinitionGroup>
     <ClCompile>
       <RuntimeTypeInfo>false</RuntimeTypeInfo>
```
@vmoroz we configure this in a few projects but not others. E.g. I think Microsoft.ReactNative and our Desktop DLL will still have RTTI enabled.

Thoughts on the idea of adding this to React.cpp.props so we can centrally control it and ensure it's enabled for all projects? It's likely some internal non-office projects including React.cpp.props would be transiently affected, but they should be moving to Microsoft.ReactNative and should not need to import that property sheet at all soon.

vmoroz replied:
We should try to remove the RTTI unless we have tests that might fail. It may be worth to search for any dynamic_cast to be sure.

---
In reply to: [370626823](https://github.com/microsoft/react-native-windows/pull/3948#discussion_r370626823) [](ancestors = 370626823)

NickGerleman commented on vnext/PropertySheets/ReactCommunity.cpp.props:
```diff
@@ -15,13 +15,7 @@
 
-  <ItemDefinitionGroup Condition="'$(OSS_RN)'=='true'">
-    <ClCompile>
-      <PreprocessorDefinitions>OSS_RN;ENUM_BITFIELDS_NOT_SUPPORTED;%(PreprocessorDefinitions)</PreprocessorDefinitions>
```
ENUM_BITFIELDS_NOT_SUPPORTED is not checked in facebook/react-native or folly. Safe to remove this, since it seems historical

NickGerleman commented on vnext/PropertySheets/React.Cpp.props:
```diff
@@ -70,15 +70,19 @@
+        BOOST_NO_RTTI; 
+        BOOST_NO_TYPEID;
+        BOOST_SYSTEM_SOURCE;
+        GTEST_HAS_RTTI=0;
```
I'd also be fine with putting this just in test projects. See pros and cons either way

vmoroz replied:
I think it is fine. Since you are changing this code, do you mind to add the WINRT_LEAN_AND_MEAN here too?

---
In reply to: [370667565](https://github.com/microsoft/react-native-windows/pull/3948#discussion_r370667565) [](ancestors = 370667565)

NickGerleman replied:
Can do

NickGerleman commented on vnext/JSI.Desktop.UnitTests/JSI.Desktop.UnitTests.vcxproj:
```diff
@@ -50,6 +50,8 @@
         wchar_t* str = L"hello";
       from compiling. -->
       <AdditionalOptions>%(AdditionalOptions) /Zc:strictStrings</AdditionalOptions>
+      <!-- JSI Unit tests come from Facebook and currently relies on dynamic_pointer_cast -->
```
nit: grammar

NickGerleman commented on vnext/PropertySheets/React.Cpp.props:
```diff
@@ -20,12 +20,12 @@
   <PropertyGroup>
     <USE_HERMES Condition="'$(USE_HERMES)' == ''">false</USE_HERMES>
-    <HERMES_Version Condition="'$(HERMES_Version)' == ''">0.1.6</HERMES_Version>
+    <HERMES_Version Condition="'$(HERMES_Version)' == ''">0.1.6</HERMES_Version>	
```
nit: extra whitespace

NickGerleman commented on vnext/PropertySheets/React.Cpp.props:
```diff
@@ -89,11 +93,4 @@
-  <ItemDefinitionGroup Condition="'$(Platform)' == 'x64' OR '$(Platform)' == 'x86' OR '$(Platform)' == 'Win32'">
-    <ClCompile>
-      <!-- MSVC macros -->
-      <PreprocessorDefinitions>WITH_OFFICE_TRACING=0;%(PreprocessorDefinitions)</PreprocessorDefinitions>
```
We don't seem to check this anywhere anymore.

vmoroz commented on vnext/ReactUWP/EndPoints/dll/react-native-uwp.x86.def:
```diff
@@ -40,9 +40,8 @@ EXPORTS
 ?JsStringToStdStringUtf8@react@facebook@@YG?AW4_JsErrorCode@@PAXAAV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@@Z
 ?LoadJavaScriptInProxyMode@DevSupportManager@uwp@react@@UAE?AV?$function@$$A6G?AV?$unique_ptr@VJSExecutor@react@facebook@@U?$default_delete@VJSExecutor@react@facebook@@@std@@@std@@V?$shared_ptr@VExecutorDelegate@react@facebook@@@2@V?$shared_ptr@VMessageQueueThread@react@facebook@@@2@@Z@std@@ABUDevSettings@3facebook@@@Z
-?LoadJavaScriptInSandboxMode@DevSupportManager@uwp@react@@UAE?AV?$function@$$A6G?AV?$unique_ptr@VJSExecutor@react@facebook@@U?$default_delete@VJSExecutor@react@facebook@@@std@@@std@@V?$shared_ptr@VExecutorDelegate@react@facebook@@@2@V?$shared_ptr@VMessageQueueThread@react@facebook@@@2@@Z@std@@ABUDevSettings@3facebook@@@Z
+;?LoadJavaScriptInSandboxMode@DevSupportManager@uwp@react@@UAE?AV?$function@$$A6G?AV?$unique_ptr@VJSExecutor@react@facebook@@U?$default_delete@VJSExecutor@react@facebook@@@std@@@std@@V?$shared_ptr@VExecutorDelegate@react@facebook@@@2@V?$shared_ptr@VMessageQueueThread@react@facebook@@@2@@Z@std@@ABUDevSettings@3facebook@@@Z
```
>;? [](start = 0, length = 2)

Is it a typo? It does not look like a valid symbol name.

vmoroz replied:
In case if it is commented out, then it is better just to remove it.

---
In reply to: [370711760](https://github.com/microsoft/react-native-windows/pull/3948#discussion_r370711760) [](ancestors = 370711760)

NickGerleman replied:
This was a rename of the file that was previously being used for OSS defs. It had it commented out since we compile it out in the non-patched version.

NickGerleman replied:
I do like how the upstream defs file commented out symbols we couldn't expose due to lacking a patch. This wasn't consistently being used though, and the exports here moved from the previous upstream defs file is definitely a bit out of data.

I'm going to make the non-patched deps follow the patched ones more closely, with them being the same except for the ones we can't yet expose commented out.

vmoroz replied:
Why to comment and not to remove?
Having commented out line could still make some developers believe that we export the symbol, and spend time debugging why something doe snot work.

---
In reply to: [371018277](https://github.com/microsoft/react-native-windows/pull/3948#discussion_r371018277) [](ancestors = 371018277)

NickGerleman replied:
I've gotten use out of them if I wanted to quickly find which of the  symbols we avoid exporting due to lack of patches. This is the approach that was used before for upstream React Native exports, and I thought the intent was relatively clear when I saw it.

vmoroz commented on vnext/PropertySheets/React.Cpp.props:
```diff
@@ -70,15 +70,19 @@
-        BOOST_SYSTEM_SOURCE;
+        BOOST_ALL_NO_LIB=1;
         BOOST_ERROR_CODE_HEADER_ONLY;
+        BOOST_NO_RTTI; 
```
>  [](start = 22, length = 1)

nit: extra space

NickGerleman commented on vnext/JSI.Desktop.UnitTests/JSI.Desktop.UnitTests.vcxproj:
```diff
@@ -50,6 +50,8 @@
         wchar_t* str = L"hello";
       from compiling. -->
       <AdditionalOptions>%(AdditionalOptions) /Zc:strictStrings</AdditionalOptions>
```
@hansenyy Does Facebook make these unit tests public in newer versions of React Native? If not we should try to get these exposed as part of their npm package.

NickGerleman commented on vnext/Folly/Folly.vcxproj:
```diff
@@ -270,7 +270,7 @@
   <Target Name="DownloadFolly" BeforeTargets="PrepareForBuild">
     <Message Importance="High" Text="Downloading folly..." Condition="!Exists('$(FollyDir)..\.follyzip\folly-2019.09.30.00.zip')" />
-    <DownloadFile Condition="!Exists('$(FollyDir)..\.follyzip\folly-2019.09.30.00.zip')" SourceUrl="https://github.com/facebook/folly/archive/v2019.09.30.00.zip" DestinationFolder="$(FollyDir)..\.follyzip" />
+    <DownloadFile Condition="!Exists('$(FollyDir)..\.follyzip\folly-2019.09.30.00.zip')" SourceUrl="https://github.com/facebook/folly/archive/v2019.09.30.00.zip" DestinationFileName="folly-2019.09.30.00.zip" DestinationFolder="$(FollyDir)..\.follyzip" />
```
This started failing when I tried to build today, since the file name on disk was v2019.09.30.00.zip. Not sure why this ever worked? We should really be doing something better here in general, like submodules.

asklar commented on vnext/ReactUWP/ReactUWP.vcxproj:
```diff
@@ -115,54 +115,30 @@
     </Link>
-    <ClCompile>
-      <RuntimeTypeInfo>false</RuntimeTypeInfo>
-    </ClCompile>
```
I'm a little skeptical about this.
To be clear: I would like us to build WITHOUT RTTI.
However from what I can see, Folly uses RTTI/dynamic_cast; if we are using Folly through source or a static library (which I think is what we do), then I think we need to have RTTI enabled (RTTI is enabled/disabled on a DLL level)

asklar replied:
(this *might* be ok if the things they're dynamic_cast-ing are never our types that we derive from theirs)

---
In reply to: [371554430](https://github.com/microsoft/react-native-windows/pull/3948#discussion_r371554430) [](ancestors = 371554430)

NickGerleman replied:
Folly looks to avoid RTTI unless FOLLY_HAS_RTTI is defined, or portability.h (which we define out) detects it's enabled by vendor specific preprocessor definitions.

NickGerleman replied:
I did a quick check and yeah, from what I can tell everywhere using a dynamic_cast is test code of experimental. typeid seems to be guarded in what I saw. If we do end up seeing folly issues we should report them since it seems like they try to guard against that definition.

NickGerleman commented:
@acoates-ms @vmoroz @asklar @kmelmon @ddalp

I removed the AutoMerge label and updated the description of this change. It should now have a lot more context, hopefully making the intent of the changes clearer. Would anyone be able to take a look at this? I'm hoping to check this in within the next day as it blocks some of the last remaining changes we know of to get off the Microsoft fork.

NickGerleman commented:
Going to go ahead and merge this. If anyone else wants to look at this, I'm open to addressing any followup concerns in a separate PR.
